File: Marmalade/MarmaladeODK/source/android/CallbacksRequestPurchase.cpp
/*
 * Copyright (C) 2012, 2013 OUYA, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "CallbacksRequestPurchase.h"
#include "CallbackSingleton.h"

#include "ODK_internal.h"
#include "s3eEdk.h"
#include "s3eEdk_android.h"
#include "IwDebug.h"

#include <stdio.h>

OuyaSDK::ExtensionProduct CallbacksRequestPurchase::m_product = OuyaSDK::ExtensionProduct();

void CallbacksRequestPurchase::RegisterCallback(s3eCallback callback, s3eCallback* savedCallback, int callbackType)
{
	if (*savedCallback)
	{
		//IwTrace(ODK, ("Unregistering Callback"));

		s3eEdkCallbacksUnRegister(
				S3E_EXT_ODK_HASH,
				S3E_ODK_CALLBACKS_MAX,
				callbackType,
				*savedCallback,
				NULL);
		*savedCallback = NULL;
	}

	*savedCallback = callback;

	//IwTrace(ODK, ("Registering Callback"));

	s3eEdkCallbacksRegister(
			S3E_EXT_ODK_HASH,
			S3E_ODK_CALLBACKS_MAX,
			callbackType,
			*savedCallback,
			NULL,
			S3E_FALSE);
}

void CallbacksRequestPurchase::RegisterCallbacks(s3eCallback onSuccess, s3eCallback onFailure, s3eCallback onCancel)
{
	RegisterCallback(onSuccess, &m_onSuccess, S3E_ODK_CALLBACKS_REQUEST_PURCHASE_ON_SUCCESS);
	RegisterCallback(onFailure, &m_onFailure, S3E_ODK_CALLBACKS_REQUEST_PURCHASE_ON_FAILURE);
	RegisterCallback(onCancel, &m_onCancel, S3E_ODK_CALLBACKS_REQUEST_PURCHASE_ON_CANCEL);
}

void CallbacksRequestPurchase::OnSuccess(const OuyaSDK::Product& product)
{
	IwTrace(ODK, ("OnSuccess"));

	OuyaSDK::ExtensionProduct eProduct;
	eProduct.Init();
	eProduct.Copy(product);
	m_product = eProduct;

	s3eRequestPurchaseSuccessEvent event;
	event.m_product = m_product;

	m_dataRequestPurchaseSuccessEvent = event; //don't send a temp pointer
	s3eEdkCallbacksEnqueue(S3E_EXT_ODK_HASH, S3E_ODK_CALLBACKS_REQUEST_PURCHASE_ON_SUCCESS, &m_dataRequestPurchaseSuccessEvent);
}

void CallbacksRequestPurchase::OnFailure(int errorCode, const std::string& errorMessage)
{
	char buffer[256];
	sprintf(buffer, "OnFailure errorCode=%d errorMessage=%s", errorCode, errorMessage.c_str());
	IwTrace(ODK, (buffer));

	s3eRequestPurchaseFailureEvent event;
	event.m_errorCode = errorCode;
	sprintf(event.m_errorMessage, "%s", errorMessage.c_str());

	m_dataRequestPurchaseFailureEvent = event; //don't send a temp pointer
	s3eEdkCallbacksEnqueue(S3E_EXT_ODK_HASH, S3E_ODK_CALLBACKS_REQUEST_PURCHASE_ON_FAILURE, &m_dataRequestPurchaseFailureEvent);
}

void CallbacksRequestPurchase::OnCancel()
{
	IwTrace(ODK, ("OnCancel"));

	s3eRequestPurchaseCancelEvent event;

	m_dataRequestPurchaseCancelEvent = event; //don't send a temp pointer
	s3eEdkCallbacksEnqueue(S3E_EXT_ODK_HASH, S3E_ODK_CALLBACKS_REQUEST_PURCHASE_ON_CANCEL, &m_dataRequestPurchaseCancelEvent);
}

File: Marmalade/MarmaladeODK/source/android/CallbacksRequestPurchase.java
/*
 * Copyright (C) 2012, 2013 OUYA, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tv.ouya.sdk.marmalade;


import android.util.Log;


public class CallbacksRequestPurchase {

	private final String LOG_TAG  = "CallbacksRequestPurchase";

	public String m_purchasable = "";

	public native void CallbacksRequestPurchaseOnSuccess(String jsonData);
	public native void CallbacksRequestPurchaseOnFailure(int errorCode, String errorMessage);
	public native void CallbacksRequestPurchaseOnCancel();

	public void onSuccess(final String jsonData) {
		Log.i("CallbacksRequestPurchase", "onSuccess jsonData=" + jsonData);
		CallbacksRequestPurchaseOnSuccess(jsonData);
	}

	public void onFailure(final int errorCode, final String errorMessage) {
		Log.i("CallbacksRequestPurchase", "onFailure: errorCode=" + errorCode + " errorMessagee=" + errorMessage);
		CallbacksRequestPurchaseOnFailure(errorCode, errorMessage);
	}

	public void onCancel() {
		Log.i("CallbacksRequestPurchase", "onCancel");
		CallbacksRequestPurchaseOnCancel();
	}
}